public/app/features/explore/slate-plugins/suggestions.tsx

```diff
@@ -186,128 +197,124 @@ export default function SuggestionsPlugin({
+  let prefix = text.slice(0, selection.focus.offset);
+
+  if (filteredDecorations.length) {
+    text = value.focusText.text.slice(filteredDecorations[0].start.offset, filteredDecorations[0].end.offset);
+    prefix = value.focusText.text.slice(filteredDecorations[0].start.offset, selection.focus.offset);
+  }
+
+  // Label values could have valid characters erased if `cleanText()` is
+  // blindly applied, which would undesirably interfere with suggestions
+  const labelValueMatch = prefix.match(/(?:!?=~?"?|")(.*)/);
+  if (labelValueMatch) {
+    prefix = labelValueMatch[1];
+  } else if (cleanText) {
+    prefix = cleanText(prefix);
+  }
+
+  const { suggestions, context } = await onTypeahead({
+    prefix,
+    text,
+    value,
+    wrapperClasses,
+    labelKey,
+  });
+
+  const filteredSuggestions = suggestions
+    .map(group => {
+      if (!group.items) {
+        return group;
+      }
 
-        if (prefix) {
-          // Filter groups based on prefix
-          if (!group.skipFilter) {
-            group.items = group.items.filter(c => (c.filterText || c.label).length >= prefix.length);
-            if (group.prefixMatch) {
-              group.items = group.items.filter(c => (c.filterText || c.label).startsWith(prefix));
-            } else {
-              group.items = group.items.filter(c => (c.filterText || c.label).includes(prefix));
-            }
+      if (prefix) {
+        // Filter groups based on prefix
+        if (!group.skipFilter) {
+          group.items = group.items.filter(c => (c.filterText || c.label).length >= prefix.length);
+          if (group.prefixMatch) {
+            group.items = group.items.filter(c => (c.filterText || c.label).startsWith(prefix));
+          } else {
+            group.items = group.items.filter(c => (c.filterText || c.label).includes(prefix));
           }
-
-          // Filter out the already typed value (prefix) unless it inserts custom text
-          group.items = group.items.filter(c => c.insertText || (c.filterText || c.label) !== prefix);
         }
 
-        if (!group.skipSort) {
-          group.items = sortBy(group.items, (item: CompletionItem) => item.sortText || item.label);
-        }
+        // Filter out the already typed value (prefix) unless it inserts custom text
+        group.items = group.items.filter(c => c.insertText || (c.filterText || c.label) !== prefix);
+      }
 
-        return group;
-      })
-      .filter(group => group.items && group.items.length); // Filter out empty groups
+      if (!group.skipSort) {
+        group.items = sortBy(group.items, (item: CompletionItem) => item.sortText || item.label);
+      }
 
-    state = {
-      ...state,
-      groupedItems: filteredSuggestions,
-      typeaheadPrefix: prefix,
-      typeaheadContext: context,
-      typeaheadText: text,
-    };
+      return group;
+    })
+    .filter(group => group.items && group.items.length); // Filter out empty groups
+
+  onStateChange({
```

Comment:
Remove reference to global state, weirdly this function is async but we do not wait for the result also I am not sure if the `editor.blur().focus();` needs to be done after state change so passing a function seemed like a more sensible solution than returning new state which would need more changes.